### PR TITLE
578 data base interfaces for plan evaluation

### PR DESCRIPTION
**Pull request by githengi**

### CODE REVIEW

**Review by ekigamba — CHANGES_REQUESTED** [review]

File: opensrp-app/src/main/java/org/smartregister/repository/dao/EventDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Created by samuelgithengi on 9/3/20.
+ */
+public class EventDaoImpl extends EventClientRepository implements EventDao {
+
+    @Override
+    public List<QuestionnaireResponse> findEventsByEntityIdAndPlan(String resourceId, String
+            planIdentifier) {
+        return fetchEvents(String.format("select %s from %s where %s =? and (%s is null or %s !=? )", event_column.json,
```
[the remark this answers is not in the document]
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/repository/dao/EventDaoImpl.java:23]: Nice catch. This is a bug, let me update should be planIdentifier is null or is equal to the plan being passed

File: opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    public List<Task> findAllTasksForEntity(String entityId) {
+        return getTasksByEntity(entityId)
+                .stream()
+                .map(TaskConverter::convertTasktoFihrResource)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public org.smartregister.domain.Task getTaskByEntityId(String taskId) {
+        return getTaskByIdentifier(taskId);
```
[the remark this answers is not in the document]
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java:74]: The interface is note properly named on plan evaluator, let me confirm and make a PR on the library

File: opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java [hunk 3]
```diff
@@ -2007,6 +2041,39 @@ public ColumnAttribute column() {
...
+    }
+
+    public boolean deleteEventsByBaseEntityId(String baseEntityId, String eventType) {
+
+        try {
+            int rowsAffected = getWritableDatabase().delete(eventTable.name(),
+                    event_column.baseEntityId.name()
+                            + " = ? AND "
+                            + event_column.eventType.name()
+                            + " != ?",
```
Comment (Rkareko) [inline opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java:2066]: A bit of clarification here, the method says delete by baseEntityId, the params are baseEntityId and eventType yet this checks that `eventType != ` to the `eventType` passed as a param.

File: opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                .map(TaskConverter::convertTasktoFihrResource)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void saveTask(org.smartregister.domain.Task task, QuestionnaireResponse questionnaireResponse) {
+        if (questionnaireResponse != null) {
+            FHIRPathElementNode structure = PathEvaluatorLibrary.getInstance()
+                    .evaluateElementExpression(questionnaireResponse,
+                            "$this.item.where(url='details' and linkId='location_id').answer");
```
Comment (Rkareko) [inline opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java:44]: Could we make this a constant with a descriptive name.

File: opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java [hunk 5]
```diff
@@ -272,14 +280,20 @@ private boolean populateStatement(SQLiteStatement statement, Table table, JSONOb
                     statement.bindString(columnOrder.get(event_column.eventId.name()), jsonObject.getString(EVENT_ID));
                 else if (jsonObject.has(_ID))
                     statement.bindString(columnOrder.get(event_column.eventId.name()), jsonObject.getString(_ID));
+                JSONObject details = jsonObject.optJSONObject(AllConstants.DETAILS);
+                if (details != null) {
+                    bindString(statement, columnOrder.get(event_column.planId.name()), details.optString(AllConstants.PLAN_IDENTIFIER));
```
[the remark this answers is not in the document]
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java:285]: apps are to do a database migration to add the column name

File: opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java (hunk 3 above)
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java:2066]: Not sure, I did not change this method

File: opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java (hunk 4 above)
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java:44]: You mean constants for the whole expression

File: opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @Override
+    public org.smartregister.domain.Task getTaskByEntityId(String taskId) {
+        return getTaskByIdentifier(taskId);
+    }
+
+    @Override
+    public void updateTask(org.smartregister.domain.Task task) {
+        addOrUpdate(task);
```
[the remark this answers is not in the document]
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java:82]: added

File: opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java [hunk 7]
```diff
@@ -259,6 +260,13 @@ private boolean populateStatement(SQLiteStatement statement, Table table, JSONOb
                 statement.bindString(columnOrder.get(client_column.syncStatus.name()), syncStatus);
                 statement.bindString(columnOrder.get(client_column.validationStatus.name()), BaseRepository.TYPE_Valid);
                 statement.bindString(columnOrder.get(client_column.baseEntityId.name()), jsonObject.getString(client_column.baseEntityId.name()));
+
+                bindString(statement, columnOrder.get(client_column.locationId.name()), jsonObject.optString(AllConstants.LOCATION_ID));
```
[the remark this answers is not in the document]
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java:264]: apps should upgrade their db when upgrading core

File: opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ekigamba) [reply opensrp-app/src/main/java/org/smartregister/repository/dao/TaskDaoImpl.java:82]: Cool, I have seen the change. Kindly update this call to pass true for an update-only operation

File: opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ekigamba) [reply opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java:285]: Cool. Have this documented on the release/tag note & announce

File: opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ekigamba) [reply opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java:264]: Cool. Have this documented on the release/tag note & announce

**Review by ekigamba — APPROVED** [review]

**Pull request merged**